### ISSUE TITLE
Decide on session middleware

### ISSUE THREAD

**Issue by untitaker** [iron/iron#495]
It appears that there are currently four (!!!) session middleware implementations available:

* https://crates.io/crates/iron_session by @dgriffen

  Appears to store session data in RAM
* https://crates.io/crates/session by @hikelee

  Redis as backend
* https://crates.io/crates/iron-sessionstorage by me

  Generic backend trait, signed cookie backend implemented. Theoretically the user could write a redis backend and plug it in
* https://github.com/iron/session, the current official one

  Generic backend trait as well, but with current API not possible to implement signed cookie backend. Default impl stores session data in RAM

We should *probably* find a consensus on those...

**Comment by JoshLambda** [comment]
**TL;DR:** Project 1 and 4 seem a bit weird in the sense that they don't seem to practically address a user base with real life use cases. Project 2 offers the most features and examples, but is being heavily tied to Redis which can be a problem. On my opinion project 3 seems practical and modular, thus the most promising regarding further evolution.

..........

A subjective view on those 4 projects:

**1** - [Daniel Griffen/iron_session](https://crates.io/crates/iron_session)

Pros:
- expiring sessions with encrypted cookies

Cons:
- abandoned project
- undocumented
- no examples

**2** - [hikelee/session](https://crates.io/crates/session)

Pros:
- expiring sessions with encrypted cookies
- has an example with login/pass auth

Cons:
- heavily tied to Redis

**3** - [untitaker/iron-sessionstorage](https://crates.io/crates/iron-sessionstorage)

Pros:
- encrypted cookies
- multiple possible backends (so far: client side with cookies, or redis)
- a basic example
- some documentation

Cons:
- no expiring sessions
- some examples missing (login/path auth, using redis)

**4** - [iron/session](https://github.com/iron/session)

Pros:
- iron official repository

Cons:
- no web-session features (just an abstract key/value datastore)
- provided example doesn't make much sense for a webapp

**Comment by untitaker** [comment]
Redis is unfinished yet in my library. It doesn't actually do anything yet

**Comment by JoshLambda** [comment]
OK, I updated my comment.

**Comment by hikelee** [comment]
Why not put the session middleware in iron crate? i think a web framework should have basic componments such as session,params,routers and so on. it waste developers' time to look around in crates.io for these basic functions.

**Comment by untitaker** [comment]
I tend to agree with you for e.g. web frameworks in Python or other higher-level languages, but keep in mind that the usecases for Rust web applications are likely to be different. If one were to write an API server, that middleware may be completely unnecessary and slow down compiletime at best, bloat binaries at worst.

**Comment by untitaker** [comment]
Also I'm trying to minimize the waste of time here by moving as much as possible into the iron org.

**Comment by JoshLambda** [comment]
Iron was created with the idea of being a micro-framework with plug-ins (see [Philosophy](https://github.com/iron/iron#philosophy)). It gives the freedom to choose plug-ins independently. As @untitaker said above there isn't to much of a loss of time if they are all referenced on the front page. The drawback I would rather see is that quite often plug-ins tend to get less development than the core. The reason for this discussion about sessions plug-ins is a rather good example.

**Comment by untitaker** [comment]
Though I probably wouldn't have anything against including params/urlencoded/body-parser into iron, as most of those are not that large.

**Comment by JoshLambda** [comment]
[This example](https://github.com/kaj/nickel-jwt-session/blob/master/examples/server-secure-multiple-routes.rs) for Nickel is really good, it would be great if we had something like that for Iron.

**Comment by untitaker** [comment]
That library is acting on a different level though. The session libraries only allow you to set key-value data on the user session, this implements an entire auth layer

**Comment by JoshLambda** [comment]
Well, this file isn't part of the library, it's an example file located in /example. 
The library itself doesn't do more than your own implementation (a key/value storage with cookies). The whole auth logic is contained in this example file only.
Still it provides good reusable code, which I think is part of framework value and great for user productivity.

**Comment by untitaker** [comment]
Not entirely sure what you're looking for, just added a barebones login example: https://github.com/untitaker/iron-sessionstorage/blob/master/examples/login.rs

**Comment by JoshLambda** [comment]
That's simply perfect. It will help me and many others with getting started.
Thanks!

**Comment by untitaker** [comment]
I moved iron-sessionstorage to the iron org but am hesitant to remove the old library. I also can't really rename it to "session" because otherwise we need a different name on crates.io.

**Comment by brunoqc** [comment]
Maybe the sessions link from the README should be removed since it now points to https://github.com/iron-graveyard/session

**Comment by phlmn** [comment]
This seems to addressed with https://github.com/iron/iron-sessionstorage.

**Comment by dgige** [comment]
I would like to know if iron-sessionstorage is still maintained?
I tried to create a new project with cargo, add iron and iron-sessionstorage in cargo.toml then copy basic.rs content into my main.rs... Compilation failed!

**Comment by phlmn** [comment]
@dgige There hasn't been an update for quite some time. Nevertheless the package will be moved to the main repository and should be considered a core package. So it definitely is still maintained. 

Can you crate an Issue for the problem here in this repository?
